README
NAME
    Text::ZPL - Encode and decode ZeroMQ Property Language

SYNOPSIS
      # Decode ZPL to a HASH:
      my $data = decode_zpl( $zpl_text );
      # Encode a HASH to ZPL text:
      my $zpl = encode_zpl( $data );

DESCRIPTION
    An implementation of the "ZeroMQ Property Language", a simple ASCII
    configuration file format; see <http://rfc.zeromq.org/spec:4> for
    details.

    Exports two functions by default: "decode_zpl" and "encode_zpl". This
    module uses Exporter::Tiny to export functions, which allows for
    flexible import options; see the Exporter::Tiny documentation for
    details.

    As a simple example, a "ZPL" file as such:

      # This is my conf.
      # There are many like it, but this one is mine.
      confname = "My Config"

      context
          iothreads = 1

      main
          publisher
              bind = tcp://eth0:5550
              bind = tcp://eth0:5551
          subscriber
              connect = tcp://192.168.0.10:5555

    ... results in a structure like:

      {
        confname => "My Config",
        context => { iothreads => '1' },
        main => {
          subscriber => {
            connect => 'tcp://192.168.0.10:5555'
          },
          publisher => {
            bind => [ 'tcp://eth0:5550', 'tcp://eth0:5551' ]
          }
        }
      }

  decode_zpl
    Given a string of "ZPL"-encoded text, returns an appropriate Perl
    "HASH"; an exception is thrown if invalid input is encountered.

    (See Text::ZPL::Stream for a streaming interface.)

  encode_zpl
    Given a Perl "HASH", returns an appropriate "ZPL"-encoded text string;
    an exception is thrown if the data given cannot be represented in "ZPL"
    (see "CAVEATS").

   TO_ZPL
    A blessed object can provide a TO_ZPL method that will supply a plain
    "HASH" or "ARRAY" (but see "CAVEATS") to the encoder:

      # Shallow-clone this object's backing hash, for example:
      sub TO_ZPL {
        my $self = shift;
        +{ %$self }
      }

  CAVEATS
    Not all Perl data structures can be represented in ZPL; specifically,
    deeply-nested structures in an "ARRAY" will throw an exception:

      # Simple list is OK:
      encode_zpl(+{ list => [ 1 .. 3 ] });
      #  -> list: 1
      #     list: 2
      #     list: 3
      # Deeply nested is not representable:
      encode_zpl(+{
        list => [
          'abc',
          list2 => [1 .. 3]
        ],
      });
      #  -> dies

    Encoding skips empty lists ("ARRAY" references).

    (The spec is unclear on all this; issues welcome via RT or GitHub!)

SEE ALSO
    Text::ZPL::Stream

AUTHOR
    Jon Portnoy <avenj@cobaltirc.org>



Changes
Revision history for Perl module Text-ZPL

0.002002 2014-10-08

  - Fix a test case on perl-5.8/5.10

0.002001 2014-10-07

  - Add Text::ZPL::Stream decoder

0.001003 2014-10-05

  - Skip empty arrays when encoding

  - Expanded documentation & tests

0.001002 2014-10-05

  - Fix handling of empty sections

0.001001 2014-10-05

  - Initial release


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Encode and decode ZeroMQ Property Language",
  "AUTHOR" => "Jon Portnoy <avenj\@cobaltirc.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Text-ZPL",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Text::ZPL",
  "PREREQ_PM" => {
    "Exporter::Tiny" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.002002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter::Tiny" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => "0.88"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Text/ZPL/Stream.pm
package Text::ZPL::Stream;
$Text::ZPL::Stream::VERSION = '0.002002';
use strict; use warnings FATAL => 'all';
use Carp;

use Text::ZPL ();



sub BUF_MAX         () { 0 }
sub BUF             () { 1 }
sub MAYBE_EXTRA_EOL () { 2 }
sub ROOT            () { 3 }
sub CURRENT         () { 4 }
sub LEVEL           () { 5 }
sub TREE            () { 6 }

sub new {
  #  max_buffer_size =>
  #  string =>
  my ($class, %param) = @_;
  my $root = +{};
  bless [
    ($param{max_buffer_size} || 0),  # BUF_MAX
    '',                              # BUF
    0,                               # MAYBE_EXTRA_EOL
    $root,                           # ROOT
    $root,                           # CURRENT
    0,                               # LEVEL
    [],                              # TREE
  ], $class
}

sub max_buffer_size {
  defined $_[0]->[BUF_MAX] ?
    $_[0]->[BUF_MAX]
    : ($_[0]->[BUF_MAX] = 0)
}


sub _maybe_extra_eol {
  $_[0]->[MAYBE_EXTRA_EOL]
}

sub _maybe_extra_eol_off {
  $_[0]->[MAYBE_EXTRA_EOL] = 0
}

sub _maybe_extra_eol_on {
  $_[0]->[MAYBE_EXTRA_EOL] = 1
}


sub _parse_current_buffer {
  my ($self) = @_;
  my $line = $self->[BUF];

  unless ( Text::ZPL::_decode_prepare_line($line) ) {
    # skippable:
    $self->[BUF] = '';
    return
  }

  Text::ZPL::_decode_handle_level(
    0, 
    $line, 
    $self->[ROOT],
    $self->[CURRENT],
    $self->[LEVEL],
    $self->[TREE],
  );
  
  if ( (my $sep_pos = index($line, '=')) > 0 ) {
    my ($k, $v) = Text::ZPL::_decode_parse_kv(
      0, $line, $self->[LEVEL], $sep_pos
    );
    Text::ZPL::_decode_add_kv(
      0, $self->[CURRENT], $k, $v
    );

    $self->[BUF] = '';
    return
  }

  my $re = $Text::ZPL::ValidName;
  if (my ($subsect) = $line =~ /^(?:\s+)?($re)(?:\s+?#.*)?$/) {
    Text::ZPL::_decode_add_subsection(
      0, $self->[CURRENT], $subsect, $self->[TREE]
    );

    $self->[BUF] = '';
    return
  }

  confess "Parse failed in ZPL stream; bad input '$line'"
}


sub get { shift->[ROOT] }

sub get_buffer { shift->[BUF] }


sub push {
  my $self = shift;
  my @chrs = split '', join '', @_;

  my $handled = 0;

  CHAR: for my $chr (@chrs) {
    if ($chr eq "\015") {
      # got \r, maybe an unneeded \n coming up, _maybe_extra_eol_on
      $self->_maybe_extra_eol_on;
      $self->_parse_current_buffer;
      ++$handled;
      next CHAR
    }
    if ($chr eq "\012") {
      if ($self->_maybe_extra_eol) {
        $self->_maybe_extra_eol_off;
      } else {
        $self->_parse_current_buffer;
        ++$handled;
      }
      next CHAR
    }

    $self->_maybe_extra_eol_off if $self->_maybe_extra_eol;

    confess "Exceeded maximum buffer size for ZPL stream"
      if  $self->max_buffer_size
      and length($self->[BUF]) >= $self->max_buffer_size;

    $self->[BUF] .= $chr
  }

  $handled
}


1;

=pod

=for Pod::Coverage BUF(_MAX)? MAYBE_EXTRA_EOL ROOT CURRENT LEVEL TREE

=head1 NAME

Text::ZPL::Stream - Streaming ZPL decoder

=head1 SYNOPSIS

  use Text::ZPL::Stream;

  my $stream = Text::ZPL::Stream->new;

  if ( $stream->push($zpl_chrs) ) {
    # Parsed at least one complete line:
    my $ref = $stream->get;
    # ...
  }

  # Or in a loop:
  while ( defined(my $zpl_chrs = magically_get_some_zpl) ) {
    $stream->push($zpl_chrs);
  }
  my $ref = $stream->get;
  # ...

=head1 DESCRIPTION

A streaming decoder for C<ZeroMQ Property Language> files using L<Text::ZPL>. 

See the L<Text::ZPL> documentation for more on C<ZPL> and parsing-related
details.

=head2 new

  my $stream = Text::ZPL::Stream->new(
    # Optional:
    max_buffer_size => 512,
  );

Constructs an object representing a new C<ZPL> stream.

Accepts the following options:

=over

=item max_buffer_size

The maximum line length allowed in buffers before an exception is thrown.

Defaults to 0 (unlimited).

=back

=head2 push

  $stream->push(@chars);
  $stream->push($string);

Takes characters (individually or as strings) and collects until an
end-of-line marker (C<\r>, C<\n>, or C<\r\n>) is encountered, at which point a
parse is called and the reference returned by L</get> is altered
appropriately.

An exception is thrown if parsing fails, or if L</max_buffer_size> is reached
-- if you're unsure of your incoming data, you may want to wrap C<push> calls
with L<Try::Tiny> or similar.

Returns the number of complete lines parsed, which can be useful as an
indicator that L</get> ought be called:

  if ( $stream->push($zpl) ) {
    # Parsed at least one complete line:
    my $ref = $stream->get;
    ...
  }

=head2 get

  my $ref = $stream->get;

Returns the C<HASH> reference to the decoded structure.

B<< This is the actual reference in use by the decoder, not a copy! >>
Altering the structure of the C<HASH> may have unintended consequences, in
which case you may want to make use of L<Storable/dclone> to create a safe
copy.

=head2 get_buffer

Returns a string containing the current character buffer (that is, any
incomplete line).

=head1 AUTHOR

Jon Portnoy <avenj@cobaltirc.org>

=cut


lib/Text/ZPL.pm
package Text::ZPL;
$Text::ZPL::VERSION = '0.002002';
use strict; use warnings FATAL => 'all';
no warnings 'void';

use Carp;
use Scalar::Util 'blessed', 'reftype';

use parent 'Exporter::Tiny';
our @EXPORT = our @EXPORT_OK = qw/
  encode_zpl
  decode_zpl
/;


# note: not anchored as-is:
our $ValidName = qr/[A-Za-z0-9\$\-_\@.&+\/]+/;


sub decode_zpl {
  my ($str) = @_;

  confess "Expected a ZPL text string but received no arguments"
    unless defined $str;

  my $root = +{};
  my $ref  = $root;
  my @descended;

  my ($level, $lineno) = (0,0);

  LINE: for my $line (split /(?:\015?\012)|\015/, $str) {
    ++$lineno;
    # Prep string in-place & skip blank/comments-only:
    next LINE unless _decode_prepare_line($line);

    # Manage structure:
    _decode_handle_level($lineno, $line, $root, $ref, $level, \@descended);

    # KV pair:
    if ( (my $sep_pos = index($line, '=')) > 0 ) {
      my ($key, $val) = _decode_parse_kv($lineno, $line, $level, $sep_pos);
      _decode_add_kv($lineno, $ref, $key, $val);
      next LINE
    }

    # New subsection:
    if (my ($subsect) = $line =~ /^(?:\s+)?($ValidName)(?:\s+?#.*)?$/) {
      _decode_add_subsection($lineno, $ref, $subsect, \@descended);
      next LINE
    }

    confess
       "Invalid ZPL (line $lineno); "
      ."unrecognized syntax or bad section name: '$line'"
  } # LINE

  $root
}

sub _decode_prepare_line {
  $_[0] =~ s/\s+$//;
  length($_[0]) == 0 || $_[0] =~ /^(?:\s+)?#/ ? () : 1
}

sub _decode_handle_level {
  # ($lineno, $line, $root, $ref, $level, $tree_ref)
  # 
  # Manage indentation-based hierarchy
  # Validates indent level
  # Munges current $ref, $level, $tree_ref in-place

  my $cur_indent = 0;
  $cur_indent++ while substr($_[1], $cur_indent, 1) eq ' ';
  if ($cur_indent % 4) {
    confess
       "Invalid ZPL (line $_[0]); "
      ."expected 4-space indent, indent is $cur_indent"
  }

  if ($cur_indent == 0) {
    $_[3] = $_[2];
    $_[4] = $cur_indent;
    @{ $_[5] } = ();
  } elsif ($cur_indent > $_[4]) {
    unless (defined $_[5]->[ ($cur_indent / 4) - 1 ]) {
      confess "Invalid ZPL (line $_[0]); no matching parent section"
    }
    $_[4] = $cur_indent;
  } elsif ($cur_indent < $_[4]) {
    my $wanted_idx = ( ($_[4] - $cur_indent) / 4 ) - 1 ;
    my $wanted_ref = $_[5]->[$wanted_idx];
    unless (defined $wanted_ref) {
      confess
        "BUG; cannot find matching parent section"
        ." [idx = $wanted_idx] [indent = $cur_indent]"
    }
    $_[3] = $wanted_ref;
    $_[4] = $cur_indent;
    @{ $_[5] } = @{ $_[5] }[ ($wanted_idx + 1) .. $#{ $_[5] } ];
  }
}

sub _decode_add_subsection {
  # ($lineno, $ref, $subsect, \@descended)
  if (exists $_[1]->{ $_[2] }) {
    confess "Invalid ZPL (line $_[0]); existing property with this name"
  }
  unshift @{ $_[3] }, $_[1];
  $_[1] = $_[1]->{ $_[2] } = +{};
}


sub _decode_parse_kv {
  # ($lineno, $line, $level, $sep_pos)
  #
  # Takes a line that appears to contain a k = v pair
  # Returns ($key, $val)

  my $key = substr $_[1], $_[2], ( $_[3] - $_[2] );
  $key =~ s/\s+$//;
  unless ($key =~ /^$ValidName$/) {
    confess "Invalid ZPL (line $_[0]); "
            ."'$key' is not a valid ZPL property name"
  }

  my $tmpval = substr $_[1], $_[3] + 1;
  $tmpval =~ s/^\s+//;
  my $realval;
  my $maybe_q = substr $tmpval, 0, 1;
  if ( ($maybe_q eq q{'} || $maybe_q eq q{"}) 
    && (my $matching_q_pos = index $tmpval, $maybe_q, 1) > 1 ) {
    # Quoted, consume up to matching and clean up tmpval
    $realval = substr $tmpval, 1, ($matching_q_pos - 1), '';
    substr $tmpval, 0, 2, '';
  } else {
    # Unquoted or mismatched quotes
    my $maybe_trailing = index $tmpval, ' ';
    $realval = substr $tmpval, 0,
      ($maybe_trailing > -1 ? $maybe_trailing : length $tmpval),
      '';
  }

  $tmpval =~ s/#.*$//;
  $tmpval =~ s/\s+//;
  if (length $tmpval) {
    confess "Invalid ZPL (line $_[0]); garbage at end-of-line: '$tmpval'"
  }

  ($key, $realval)
}

sub _decode_add_kv {
  # ($lineno, $ref, $key, $val)
  #
  # Add a value to property; create lists as-needed

  if (exists $_[1]->{ $_[2] }) {
    if (ref $_[1]->{ $_[2] } eq 'HASH') {
      confess
        "Invalid ZPL (line $_[0]); existing subsection with this name"
    } elsif (ref $_[1]->{ $_[2] } eq 'ARRAY') {
      push @{ $_[1]->{ $_[2] } }, $_[3]
    } else {
      $_[1]->{ $_[2] } = [ $_[1]->{ $_[2] }, $_[3] ]
    }
    return
  }
  $_[1]->{ $_[2] } = $_[3]
}


sub encode_zpl {
  my ($obj) = @_;
  $obj = $obj->TO_ZPL if blessed $obj and $obj->can('TO_ZPL');
  confess "Expected a HASH but got $obj" unless ref $obj eq 'HASH';
  _encode($obj)
}

sub _encode {
  my ($ref, $indent) = @_;
  $indent ||= 0;
  my $str = '';

  KEY: for my $key (keys %$ref) {
    confess "$key is not a valid ZPL property name"
      unless $key =~ qr/^$ValidName$/;
    my $val = $ref->{$key};
    
    if (blessed $val && $val->can('TO_ZPL')) {
      $val = $val->TO_ZPL;
    }

    if (ref $val eq 'ARRAY') {
      $str .= _encode_array($key, $val, $indent);
      next KEY
    }

    if (ref $val eq 'HASH') {
      $str .= ' ' x $indent;
      $str .= "$key\n";
      $str .= _encode($val, $indent + 4);
      next KEY
    }
    
    if (ref $val) {
      confess "Do not know how to handle '$val'"
    }

    $str .= ' ' x $indent;
    $str .= "$key = " . _maybe_quote($val) . "\n";
  }

  $str
}

sub _encode_array {
  my ($key, $ref, $indent) = @_;
  my $str = '';
  for my $item (@$ref) {
    confess "ZPL does not support structures of this type in lists: ".ref $item
      if ref $item;
    $str .= ' ' x $indent;
    $str .= "$key = " . _maybe_quote($item) . "\n";
  }
  $str
}

sub _maybe_quote {
  my ($val) = @_;
  return qq{'$val'}
    if index($val, q{"}) > -1
    and index($val, q{'}) == -1;
  return qq{"$val"}
    if index($val, '#')  > -1
    or index($val, '=')  > -1
    or (index($val, q{'}) > -1 and index($val, q{"}) == -1)
    or $val =~ /\s/;  # last because slow :\
  $val
}

1;

=pod

=head1 NAME

Text::ZPL - Encode and decode ZeroMQ Property Language

=head1 SYNOPSIS

  # Decode ZPL to a HASH:
  my $data = decode_zpl( $zpl_text );
  # Encode a HASH to ZPL text:
  my $zpl = encode_zpl( $data );

=head1 DESCRIPTION

An implementation of the C<ZeroMQ Property Language>, a simple ASCII
configuration file format; see L<http://rfc.zeromq.org/spec:4> for details.

Exports two functions by default: L</decode_zpl> and L</encode_zpl>. This
module uses L<Exporter::Tiny> to export functions, which allows for flexible
import options; see the L<Exporter::Tiny> documentation for details.

As a simple example, a C<ZPL> file as such:

  # This is my conf.
  # There are many like it, but this one is mine.
  confname = "My Config"

  context
      iothreads = 1

  main
      publisher
          bind = tcp://eth0:5550
          bind = tcp://eth0:5551
      subscriber
          connect = tcp://192.168.0.10:5555

... results in a structure like:

  {
    confname => "My Config",
    context => { iothreads => '1' },
    main => {
      subscriber => {
        connect => 'tcp://192.168.0.10:5555'
      },
      publisher => {
        bind => [ 'tcp://eth0:5550', 'tcp://eth0:5551' ]
      }
    }
  }

=head2 decode_zpl

Given a string of C<ZPL>-encoded text, returns an appropriate Perl C<HASH>; an
exception is thrown if invalid input is encountered.

(See L<Text::ZPL::Stream> for a streaming interface.)

=head2 encode_zpl

Given a Perl C<HASH>, returns an appropriate C<ZPL>-encoded text string; an
exception is thrown if the data given cannot be represented in C<ZPL> (see
L</CAVEATS>).

=head3 TO_ZPL

A blessed object can provide a B<TO_ZPL> method that will supply a plain
C<HASH> or C<ARRAY> (but see L</CAVEATS>) to the encoder:

  # Shallow-clone this object's backing hash, for example:
  sub TO_ZPL {
    my $self = shift;
    +{ %$self }
  }

=head2 CAVEATS

Not all Perl data structures can be represented in ZPL; specifically,
deeply-nested structures in an C<ARRAY> will throw an exception:

  # Simple list is OK:
  encode_zpl(+{ list => [ 1 .. 3 ] });
  #  -> list: 1
  #     list: 2
  #     list: 3
  # Deeply nested is not representable:
  encode_zpl(+{
    list => [
      'abc',
      list2 => [1 .. 3]
    ],
  });
  #  -> dies

Encoding skips empty lists (C<ARRAY> references).

(The spec is unclear on all this; issues welcome via RT or GitHub!)

=head1 SEE ALSO

L<Text::ZPL::Stream>

=head1 AUTHOR

Jon Portnoy <avenj@cobaltirc.org>

=cut
